### ISSUE TITLE
20x slowdown recording Firefox `SandboxBrokerTest.MultiThread*` unit tests

### ISSUE THREAD

**Issue by jld** [mozilla/rr#2159]
The Firefox GTests `SandboxBrokerTest.MultiThreadOpen` and `SandboxBrokerTest.MultiThreadStat` run 20x slower than native (constrained to one core) when being recorded with rr 5.1.0.

These are unit tests of the Linux sandbox file broker; seccomp-bpf is not involved.  They construct a broker instance and send it a lot of requests from several threads at once in a tight loop, within the same process.  (And they're the least unreliable reproducer we have for a [mysterious intermittent bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1243108).)  For each request, a `socketpair` is created and one end is sent to the broker with `SCM_RIGHTS` to carry the response.

I'm not surprised that rr has to do nontrivial work for this, and it's probably not something a normal application would do on the critical path, but this is a lot of slowdown.

For reference, the command I've been running is `MOZ_RUN_GTEST=True GTEST_FILTER=SandboxBrokerTest.MultiThread\* rr record obj-x86_64-pc-linux-gnu/dist/bin/firefox -unittest`, and I'm using a non-debug build.

**Comment by rocallahan** [comment]
I assume this is basically a ping-pong-between processes test where there are lots of context switches and very little per-process work between context switches, right?
And it also sounds like it runs multicore without rr and rr forces it onto a single core.
So I would expect this to be a worst-case scenario for rr recording overhead.

Can you run it on a single core outside of rr using `taskset` and report what the overhead is for just forcing everything onto a single core?

**Comment by jld** [comment]
Yes, it's mostly context switches (normally between processes, but in this unit test it's between threads in one process).

Normal:
```
[       OK ] SandboxBrokerTest.MultiThreadOpen (712 ms)
[       OK ] SandboxBrokerTest.MultiThreadStat (871 ms)
[----------] 2 tests from SandboxBrokerTest (1583 ms total)
```

Single core:
```
[       OK ] SandboxBrokerTest.MultiThreadOpen (1408 ms)
[       OK ] SandboxBrokerTest.MultiThreadStat (1492 ms)
[----------] 2 tests from SandboxBrokerTest (2900 ms total)
```

rr:
```
[       OK ] SandboxBrokerTest.MultiThreadOpen (27089 ms)
[       OK ] SandboxBrokerTest.MultiThreadStat (24863 ms)
[----------] 2 tests from SandboxBrokerTest (51953 ms total)
```

rr chaos mode:
```
[       OK ] SandboxBrokerTest.MultiThreadOpen (32537 ms)
[       OK ] SandboxBrokerTest.MultiThreadStat (33792 ms)
[----------] 2 tests from SandboxBrokerTest (66330 ms total)
```

**Comment by rocallahan** [comment]
Unfortunately there's probably not much we can do about this. We could probably do some microoptimizations to make it a bit faster, but not a lot faster.
